src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -658,6 +668,55 @@ RegenerateTaskForFasthPathQuery(Job *workerJob)
 }
 
 
+/*
+ * RePlanNonFastPathQuery rebuilds workerJob by replanning UPDATE/DELETE/SELECT
+ * non-fast path queries. It creates rebuilt job after filtering out tasks for non-local
+ * nodes.
+ */
+static void
+RePlanNonFastPathQuery(DistributedPlan **distributedPlan)
+{
+	/* replan workerquery */
+	Query *workerQuery = (*distributedPlan)->workerJob->jobQuery;
+	PlannedStmt *plannedStmt = planner(workerQuery, NULL, 0, NULL);
```

Comment:
Calling the planner here pulls in a lot of complexity, which will only get exercised in rare edge cases.

One question is: how sure are we that we can simply pass jobQuery to the planner and does not contain any citus_extradata_container calls?

Do we need to call the planner from the top? I'm wondering if there are lower level functions we can all that do a lot less work. However, that probably requires a refactor of the plan data structure that gets passed into the executor.

If we were only solving this specific problem, I would probably push back on it because the gain is not worth the complexity. However, we have similar problems around prepared statements, which already involve a lot of complexity, so I still find solving this problem quite interesting.

Reply:
Yes, we cannot guarentee if workerquery has any `citus_extradata_container`, which makes current solution buggy. When I think once, I had 2 options that I am not so sure about. 

1. I wonder if we can encounter with any problem if we first replace `citus_extradata_container` with shell table rtes and then replan the workerquery,
2. Maybe easier, but again not so sure if that would work: store originalquery inside DistirubtedPlan so that we can simply replan and replace scanstate's distributedplan.

Reply:
I added code to realize my 2. idea. It seems like it worked. I try to run multi-1 schedule with always-replan-nonfastpath code. It passes all the tests. I found out that we currently cannot support replanning modifying ctes (they are subplan so their parent plan should be updated properly). Also we cannot replan delegated functions which has NULL replanning context on the delegated node. 

I will run other test suites with the same approach (always-replan nonfast) to see if we should consider any other thing.

Reply:
Because our replan logic is generic, we can also replan fastpath queries instead of rerouting them for the sake of cleaner code. Missing shards are rare case so it should be ok to replan for all types of queries.